File: apps/dashboard/src/types/api/market-stream.ts
import { z } from 'zod';

/**
 * Wire schemas for `/api/v1/ws/market` (see `services/api/app/api/v1/endpoints/market_stream.py`
 * for the authoritative protocol description). Every inbound frame is
 * validated against these before the app trusts it, matching the same
 * convention as the REST client in `src/lib/api/market.ts`.
 */

export const LiveTradeDataSchema = z.object({
  price: z.string(),
  size: z.string(),
  side: z.string(),
  event_time: z.string().datetime(),
});

export type LiveTradeData = z.infer<typeof LiveTradeDataSchema>;

export const LiveTickerDataSchema = z.object({
  last_price: z.string().nullable(),
  bid: z.string().nullable(),
  ask: z.string().nullable(),
  mark_price: z.string().nullable(),
  price_change_24h: z.string().nullable(),
  event_time: z.string().datetime(),
});

export type LiveTickerData = z.infer<typeof LiveTickerDataSchema>;

export const OrderBookLevelSchema = z.object({
  price: z.string(),
  size: z.string(),
});

export type OrderBookLevelData = z.infer<typeof OrderBookLevelSchema>;

/**
 * Already sorted (bids descending, asks ascending) and depth-limited by the
 * gateway — see `services/api/app/marketdata/gateway.py`'s
 * `_ORDERBOOK_DEPTH` — and reconstructed from a snapshot + merged
 * incremental diffs (`app/marketdata/orderbook.py`'s `OrderBookAggregator`),
 * never a single raw exchange message.
 */
export const LiveOrderBookDataSchema = z.object({
  bids: z.array(OrderBookLevelSchema),
  asks: z.array(OrderBookLevelSchema),
  event_time: z.string().datetime().nullable(),
  sequence: z.number().int().nullable(),
});

export type LiveOrderBookData = z.infer<typeof LiveOrderBookDataSchema>;

export const MarketStreamSnapshotSchema = z.object({
  type: z.literal('snapshot'),
  symbol: z.string(),
  trade: LiveTradeDataSchema.nullable(),
  ticker: LiveTickerDataSchema.nullable(),
  orderbook: LiveOrderBookDataSchema.nullable(),
});

export const MarketStreamTradeSchema = z.object({
  type: z.literal('trade'),
  symbol: z.string(),
  data: LiveTradeDataSchema,
});

export const MarketStreamTickerSchema = z.object({
  type: z.literal('ticker'),
  symbol: z.string(),
  data: LiveTickerDataSchema,
});

export const MarketStreamOrderBookSchema = z.object({
  type: z.literal('orderbook'),
  symbol: z.string(),
  data: LiveOrderBookDataSchema,
});

export const MarketStreamPongSchema = z.object({
  type: z.literal('pong'),
});

export const MarketStreamErrorSchema = z.object({
  type: z.literal('error'),
  detail: z.string(),
});

export const MarketStreamMessageSchema = z.discriminatedUnion('type', [
  MarketStreamSnapshotSchema,
  MarketStreamTradeSchema,
  MarketStreamTickerSchema,
  MarketStreamOrderBookSchema,
  MarketStreamPongSchema,
  MarketStreamErrorSchema,
]);

export type MarketStreamMessage = z.infer<typeof MarketStreamMessageSchema>;

File: services/api/app/api/v1/endpoints/market_stream.py
"""Live market data WebSocket gateway for browser clients.

This is the only real-time channel the dashboard uses: it relays trade,
ticker, and reconstructed order-book events already flowing through the
in-process event bus (published by the Delta WebSocket client +
processing pipeline, see ``app.runtime``) to subscribed browser
connections. The frontend never opens a connection to the exchange
itself.

Protocol (JSON text frames):

Client -> server:
    ``{"action": "subscribe", "symbols": ["ETHUSD"]}``
    ``{"action": "unsubscribe", "symbols": ["ETHUSD"]}``
    ``{"action": "ping"}``

Server -> client:
    ``{"type": "snapshot", "symbol": "ETHUSD",
    "trade": {...}|null, "ticker": {...}|null, "orderbook": {...}|null}``
    ``{"type": "trade", "symbol": "ETHUSD", "data": {...}}``
    ``{"type": "ticker", "symbol": "ETHUSD", "data": {...}}``
    ``{"type": "orderbook", "symbol": "ETHUSD",
    "data": {"bids": [...], "asks": [...], "event_time", "sequence"}}``
    ``{"type": "pong"}``
    ``{"type": "error", "detail": "..."}``

``orderbook.bids``/``asks`` are already sorted (bids descending, asks
ascending) and depth-limited by the gateway — see
``app.marketdata.gateway``'s ``_ORDERBOOK_DEPTH`` — and reflect a
*reconstructed* book (snapshot + merged incremental diffs via
``app.marketdata.orderbook.OrderBookAggregator``), not a single raw
exchange message.

No authentication, no trading, no order *execution* data — this is a
read-only market-data relay (order *book* depth included), matching the
rest of the v1 API surface.
"""

import asyncio
import json
import logging
from typing import Annotated, Any

from fastapi import APIRouter, Depends, WebSocket, WebSocketDisconnect
from pydantic import BaseModel, ValidationError

from app.marketdata.gateway import ConnectionHandle
from app.runtime import Runtime, get_runtime

logger = logging.getLogger("app.api.market_stream")

router = APIRouter(tags=["market-stream"])

RuntimeDep = Annotated[Runtime, Depends(get_runtime)]

_MAX_SYMBOLS_PER_MESSAGE = 20
_KNOWN_ACTIONS = frozenset({"subscribe", "unsubscribe", "ping"})


class ClientMessage(BaseModel):
    """One inbound control message."""

    action: str
    symbols: list[str] = []


async def _writer_loop(websocket: WebSocket, handle: ConnectionHandle) -> None:
    """Drain ``handle``'s queue and send each message to the client.

    A dedicated writer task per connection is what lets the event-bus
    fan-out (a different task) enqueue messages safely — only this loop
    ever calls ``websocket.send_json``, so concurrent sends can never
    interleave on the wire.
    """
    while True:
        message = await handle.queue.get()
        await websocket.send_json(message)


async def _handle_client_message(
    runtime: Runtime, handle: ConnectionHandle, websocket: WebSocket, raw: Any
) -> None:
    try:
        message = ClientMessage.model_validate(raw)
    except ValidationError as exc:
        await websocket.send_json({"type": "error", "detail": str(exc)})
        return

    if message.action not in _KNOWN_ACTIONS:
        await websocket.send_json({"type": "error", "detail": f"unknown action '{message.action}'"})
        return

    symbols = message.symbols[:_MAX_SYMBOLS_PER_MESSAGE]
    if message.action == "subscribe":
        runtime.gateway.subscribe(handle, symbols)
    elif message.action == "unsubscribe":
        runtime.gateway.unsubscribe(handle, symbols)
    elif message.action == "ping":
        await websocket.send_json({"type": "pong"})


@router.websocket("/ws/market")
async def market_stream(websocket: WebSocket, runtime: RuntimeDep) -> None:
    """Accept a browser connection and relay live market data on demand.

    A client must send a ``subscribe`` message to receive anything; no
    symbol is implicitly subscribed on connect.
    """
    await websocket.accept()
    handle = runtime.gateway.register()
    writer_task = asyncio.create_task(_writer_loop(websocket, handle))
    try:
        while True:
            raw_text = await websocket.receive_text()
            try:
                raw = json.loads(raw_text)
            except json.JSONDecodeError:
                await websocket.send_json({"type": "error", "detail": "invalid JSON"})
                continue
            await _handle_client_message(runtime, handle, websocket, raw)
    except WebSocketDisconnect:
        logger.debug("market_stream client disconnected")
    except Exception:
        logger.exception("market_stream connection failed")
    finally:
        writer_task.cancel()
        runtime.gateway.unregister(handle)
